### PR TITLE
Recursive Calcite sub-schema support

### PR DESCRIPTION
**Pull request by GavinRay97**
Doesn't work currently, but also doesn't break existing code
I'm not sure how to create a new ML relation type and attach the results nested under that, but this is a decent scaffold of the logic (I hope)

### CODE REVIEW

File: src/main/java/net/hydromatic/morel/foreign/CalciteForeignValue.java [hunk 1]
```diff
@@ -85,24 +85,45 @@ private static String convert(boolean lower, String name) {
...
+    SchemaPlus currentSchema = schema;
+    while (true) {
+      final Set<String> subSchemaNames = requireNonNull(currentSchema).getSubSchemaNames();
+      if (subSchemaNames.isEmpty()) {
+        break;
+      }
+      for (String subSchemaName : subSchemaNames) {
+        final SchemaPlus subSchema = requireNonNull(currentSchema.getSubSchema(subSchemaName));
+        System.out.println("CalciteForeignValue - processing sub-schema: " + subSchemaName);
+        // TODO: How to create a new ML record type for each sub-schema?
```
Comment (julianhyde) [inline src/main/java/net/hydromatic/morel/foreign/CalciteForeignValue.java:101]: You need to create a record type, and an instance of it. 

`TypeSystem.recordType(SortedMap)` creates the type. One field per sub-schema (or table) in the schema.

A record value is just a Java List. The entries correspond to the fields (which are always sorted by name). You should wrap each `Schema` object as a `CalciteForeignValue`, and each `Table` object as a `RelList`. (`CalciteForeignValue.value()` seems to do a lot of this already.)

Comment (julianhyde) [reply src/main/java/net/hydromatic/morel/foreign/CalciteForeignValue.java:101]: If you want an example of code creating an instance of a record, see how `Codes.TupleCode.eval(EvalEnv)` calls `Arrays.asList`. Tuples and records are the same at runtime (the only difference is that tuples have field names "1", "2", "3" etc whereas records have alphabetical names) so this piece of code is used for both.

Comment (GavinRay97) [reply src/main/java/net/hydromatic/morel/foreign/CalciteForeignValue.java:101]: Ah okay great, thank you! Will try to implement this based on your info here 👍

Comment (GavinRay97) [reply src/main/java/net/hydromatic/morel/foreign/CalciteForeignValue.java:101]: Hmm, I think I'm close but something here is breaking it:

```java
java.lang.ClassCastException: class java.lang.Integer cannot be cast to class java.util.List (java.lang.Integer and java.util.List are in module java.base of loader 'bootstrap')
```

```java
    public Type type(TypeSystem typeSystem) {
        return type(schema, typeSystem);
    }

    // New function (recursive)
    public Type type(SchemaPlus schema, TypeSystem typeSystem) {
        final ImmutableSortedMap.Builder<String, Type> fields =
                ImmutableSortedMap.orderedBy(RecordType.ORDERING);

        schema.getTableNames().forEach(tableName ->
                fields.put(convert(tableName),
                        toType(schema.getTable(tableName), typeSystem)));

        schema.getSubSchemaNames().forEach(subSchemaName -> {
            final SchemaPlus subSchema = schema.getSubSchema(subSchemaName);
            final ImmutableSortedMap.Builder<String, Type> subFields =
                    ImmutableSortedMap.orderedBy(RecordType.ORDERING);

            subSchema.getTableNames().forEach(tableName ->
                    subFields.put(convert(tableName),
                            toType(subSchema.getTable(tableName), typeSystem)));

            subSchema.getSubSchemaNames().forEach(subSubSchemaName -> {
              subFields.put(convert(subSubSchemaName),
                      type(subSchema.getSubSchema(subSubSchemaName), typeSystem));
            });

            fields.put(subSchemaName,
                    typeSystem.recordType(subFields.build()));
        });

        return typeSystem.recordType(fields.build());
    }
```

Comment (GavinRay97) [reply src/main/java/net/hydromatic/morel/foreign/CalciteForeignValue.java:101]: I got it!
```
fieldMap = {todos={completed:bool, name:string} list}
fieldMap = {todos={completed:bool, name:string} list, users={age:int, name:string} list}
Out: val it = <relation> : {completed:bool, name:string} list
```

**Comment by GavinRay97** [comment]
Updated with test, ready for review

**Comment by julianhyde** [comment]
Thanks for the PR. In https://github.com/julianhyde/morel/tree/147-recursive-sub-schema I've fixed the formatting issues that were causing checkstyle to fail and added several other review comments.

**Comment by julianhyde** [comment]
Merged. Thanks for the PR, @GavinRay97!

**Pull request closed without merging**